### PR TITLE
tumpa: init at 0.1.1

### PR DESCRIPTION
**Pull request by 0x4A6F**
###### Motivation for this change
Add package [tumpa](https://github.com/kushaldas/tumpa) with dependency [johnnycanencrypt](https://github.com/kushaldas/johnnycanencrypt).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/tumpa/default.nix [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  propagatedBuildInputs = [
+    setuptools
+    johnnycanencrypt
+    pyside2
+  ];
+
+  meta = with stdenv.lib; {
+    description = "OpenPGP key creation and smartcard access";
+    homepage = "https://github.com/kushaldas/tumpa";
+    license = licenses.gpl3Plus;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/tumpa/default.nix:29]: Missing maintainer.

File: pkgs/development/python-modules/tumpa/default.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    rev = "v${version}";
+    sha256 = "1wvs64s0jxn4p8zr643d2hcczw3a175r6ib3481gdhjx38kgxjbq";
+  };
+
+  propagatedBuildInputs = [
+    setuptools
+    johnnycanencrypt
+    pyside2
+  ];
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/tumpa/default.nix:25]: Please add a pythonImportsCheck.

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -3180,6 +3180,8 @@ in {
 
   joblib = callPackage ../development/python-modules/joblib { };
 
+  johnnycanencrypt = disabledIf (!pythonAtLeast "3.7") (callPackage ../development/python-modules/johnnycanencrypt { });
```
Comment (SuperSandro2000) [inline pkgs/top-level/python-packages.nix:3183]: Please use disabled in the package.

File: pkgs/top-level/python-packages.nix [hunk 4]
```diff
@@ -7556,6 +7558,8 @@ in {
 
   tubeup = callPackage ../development/python-modules/tubeup { };
 
+  tumpa = disabledIf (!pythonAtLeast "3.7") (callPackage ../development/python-modules/tumpa { });
```
Comment (SuperSandro2000) [inline pkgs/top-level/python-packages.nix:7561]: Please use disabled in the package. Also there is pythonOlder.

File: pkgs/development/python-modules/johnnycanencrypt/default.nix [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  doInstallCheck = true;
+
+  installCheckInputs = [
+    pytestCheckHook
+    numpy
+  ];
+
+  buildPhase = ''
+    runHook preBuild
+    maturin build --release --manylinux off --strip
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/johnnycanencrypt/default.nix:62]:
```suggestion
    maturin build --release --manylinux off --strip --cargo-extra-args="-j $NIX_BUILD_CORES --frozen"
```

File: pkgs/development/python-modules/johnnycanencrypt/default.nix [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    sha256 = "192wfrlyylrpzq70yki421mi1smk8q2cyki2a1d03q7h6apib3j4";
+  };
+  cargoPatches = [ ./Cargo.lock.patch ];
+
+  cargoSha256 = "0ifvpdizcdp2c5x2x2j1bhhy5a75q0pk7a63dmh52mlpmh45fy6r";
+
+  LIBCLANG_PATH = llvmPackages.libclang + "/lib";
+
+  propagatedBuildInputs = [
+    python
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/johnnycanencrypt/default.nix:34]: That is a nativeBuildInput of maturin to check if python is available, right?

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
If you take a look at https://github.com/NixOS/nixpkgs/pull/108378/files#diff-a6ddec54f874bafe1f87a5423f6135222cb002ff0f8e4d0e532433aaaf3cd3c0 you will find a version which supports multiple python versions. Also I am working on a hook which does all of this. When I am done I am going to convert this package, too.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 108522` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
6 packages built:
</summary>
<ul>
<li>python37Packages.johnnycanencrypt</li>
<li>python37Packages.tumpa</li>
<li>python38Packages.johnnycanencrypt</li>
<li>python38Packages.tumpa</li>
<li>python39Packages.johnnycanencrypt</li>
<li>python39Packages.tumpa</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

python37Packages.tumpa:
<details>

Zero tests run by pytest got detected: 'Ran 0 tests in 0.000s'
You have the following options to fix this:
  - Tell pytest(CheckHook) where to find the tests included in the package
  - Check if the GitHub Repo contains tests but they are not shipped with Pypi. If so please switch to `fetchFromGitHub`.
  - If the Packages does not contain any tests add \'doCheck = false;'` and a `pythonImportsCheck`.

</details>
python38Packages.tumpa:
<details>

Zero tests run by pytest got detected: 'Ran 0 tests in 0.000s'
You have the following options to fix this:
  - Tell pytest(CheckHook) where to find the tests included in the package
  - Check if the GitHub Repo contains tests but they are not shipped with Pypi. If so please switch to `fetchFromGitHub`.
  - If the Packages does not contain any tests add \'doCheck = false;'` and a `pythonImportsCheck`.

</details>
python39Packages.tumpa:
<details>

Zero tests run by pytest got detected: 'Ran 0 tests in 0.000s'
You have the following options to fix this:
  - Tell pytest(CheckHook) where to find the tests included in the package
  - Check if the GitHub Repo contains tests but they are not shipped with Pypi. If so please switch to `fetchFromGitHub`.
  - If the Packages does not contain any tests add \'doCheck = false;'` and a `pythonImportsCheck`.

</details>

File: pkgs/development/python-modules/tumpa/default.nix (hunk 2 above)
Comment (0x4A6F) [reply pkgs/development/python-modules/tumpa/default.nix:27]: Added for johnnycanencrypt, tumpa seems to be lacking tests.

File: pkgs/development/python-modules/tumpa/default.nix (hunk 1 above)
Comment (0x4A6F) [reply pkgs/development/python-modules/tumpa/default.nix:33]: Fixed.

File: pkgs/development/python-modules/johnnycanencrypt/default.nix [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    install -Dm644 -t dist target/wheels/*.whl
+    pipInstallPhase
+  '';
+
+  preCheck = ''
+    export TESTDIR=$(mktemp -d)
+    cp -r tests/ $TESTDIR
+    pushd $TESTDIR
+  '';
+  postCheck = "popd";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/johnnycanencrypt/default.nix:78]:
```suggestion
  '';

  postCheck = ''
    popd
  '';
```

File: pkgs/development/python-modules/tumpa/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/tumpa/default.nix:27]: pythonImportsCheck is a replacement for tests to know if the module can be imported at all. Please add this.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 108522` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
6 packages built:
</summary>
<ul>
<li>python37Packages.johnnycanencrypt</li>
<li>python37Packages.tumpa</li>
<li>python38Packages.johnnycanencrypt</li>
<li>python38Packages.tumpa</li>
<li>python39Packages.johnnycanencrypt</li>
<li>python39Packages.tumpa</li>
</ul>
</details>

**Comment by SuperSandro2000** [comment]
darwin diff for johnny

```diff
diff --git a/pkgs/development/python-modules/johnnycanencrypt/default.nix b/pkgs/development/python-modules/johnnycanencrypt/default.nix
index 607c36c8da7..407bb44d8b2 100644
--- a/pkgs/development/python-modules/johnnycanencrypt/default.nix
+++ b/pkgs/development/python-modules/johnnycanencrypt/default.nix
@@ -13,6 +13,7 @@
 , numpy
 , pytestCheckHook
 , pythonOlder
+, PCSC
 }:
 
 rustPlatform.buildRustPackage rec {
@@ -48,7 +49,7 @@ rustPlatform.buildRustPackage rec {
   buildInputs = [
     pcsclite
     nettle
-  ];
+  ] ++ stdenv.lib.optionals stdenv.isDarwin [ PCSC ];
 
   # Needed b/c need to check AFTER python wheel is installed (using Rust Build, not buildPythonPackage)
   doCheck = false;
diff --git a/pkgs/top-level/python-packages.nix b/pkgs/top-level/python-packages.nix
index 75754ba26e0..935155c112d 100644
--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -3214,7 +3214,9 @@ in {
 
   joblib = callPackage ../development/python-modules/joblib { };
 
-  johnnycanencrypt = callPackage ../development/python-modules/johnnycanencrypt { };
+  johnnycanencrypt = callPackage ../development/python-modules/johnnycanencrypt {
+    inherit (pkgs.darwin.apple_sdk.frameworks) PCSC;
+  };
 
   josepy = callPackage ../development/python-modules/josepy { };
 
```

**Pull request merged**